overlord/snapstate/handlers_components.go

```diff
@@ -0,0 +1,229 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/snapstate/backend"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snapdir"
+	"github.com/snapcore/snapd/timings"
+	"gopkg.in/tomb.v2"
+)
+
+// TaskComponentSetup returns the ComponentSetup and SnapSetup with task params hold
+// by or referred to by the task.
+func TaskComponentSetup(t *state.Task) (*ComponentSetup, *SnapSetup, error) {
+	snapsup, err := TaskSnapSetup(t)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var compSetup ComponentSetup
+	err = t.Get("component-setup", &compSetup)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, nil, err
+	}
+	if err == nil {
+		return &compSetup, snapsup, nil
+	}
+
+	var id string
+	err = t.Get("component-setup-task", &id)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ts := t.State().Task(id)
+	if ts == nil {
+		return nil, nil, fmt.Errorf("internal error: tasks are being pruned")
+	}
+	if err := ts.Get("component-setup", &compSetup); err != nil {
+		return nil, nil, err
+	}
+	return &compSetup, snapsup, nil
```

Comment:
atm tests don't reach this part of the code, not sure is a big issue, as they will at some point? but pointing it out